providers/implementations/ciphers/cipher_aes_cts_fips.c

```diff
@@ -0,0 +1,313 @@
+    len -= residue;
+
+    if (!ctx->hw->cipher(ctx, out, in, len))
+        return 0;
+
+    in += len;
+    out += len;
+
+    memset(tmp_in.c, 0, sizeof(tmp_in));
+    memcpy(tmp_in.c, in, residue);
+    memcpy(out, out - 16, residue);
+    if (!ctx->hw->cipher(ctx, out - 16, tmp_in.c, 16))
+        return 0;
+    return len + residue;
+}
+
+static size_t cts128_cs3_decrypt(PROV_CIPHER_CTX *ctx, const unsigned char *in,
+                                 unsigned char *out, size_t len)
+{
+    aligned_32bytes tmp_out;
+    size_t residue;
+
+    if (len <= 16) /* CS3 requires 2 blocks */
+        return 0;
+
+    residue = len % 16;
+    if (residue == 0)
+        residue = 16;
+    len -= 16 + residue;
+
+    if (len) {
+        if (!ctx->hw->cipher(ctx, out, in, len))
+            return 0;
+        in += len;
+        out += len;
+    }
+
+    memset(tmp_out.c, 0, sizeof(tmp_out));
+    memcpy(ctx->buf, ctx->iv, 16);   /* save the iv */
+    memset(ctx->iv, 0, 16);          /* use an iv of zero */
+    if (!ctx->hw->cipher(ctx, tmp_out.c, in, 16))
+        return 0;
+    memcpy(tmp_out.c + 16, ctx->iv, 16);
+    memcpy(tmp_out.c, in + 16, residue);
+
+    memcpy(ctx->iv, ctx->buf, 16);   /* restore the iv */
+    if (!ctx->hw->cipher(ctx, tmp_out.c, tmp_out.c, 32))
+        return 0;
+
+    memcpy(out, tmp_out.c, 16 + residue);
+    return 16 + len + residue;
+}
+
+static size_t cts128_cs2_encrypt(PROV_CIPHER_CTX *ctx, const unsigned char *in,
+                                 unsigned char *out, size_t len)
+{
+    if (len % 16 == 0) {
+        /* If there are no partial blocks then it is the same as CBC mode */
+        if (!ctx->hw->cipher(ctx, out, in, len))
+            return 0;
+        return len;
+    }
+    /* For partial blocks CS2 is equivalent to CS3 */
+    return cts128_cs3_encrypt(ctx, in, out, len);
+}
+
+static size_t cts128_cs2_decrypt(PROV_CIPHER_CTX *ctx, const unsigned char *in,
+                                 unsigned char *out, size_t len)
+{
+    if (len % 16 == 0) {
+        /* If there are no partial blocks then it is the same as CBC mode */
+        if (!ctx->hw->cipher(ctx, out, in, len))
+            return 0;
+        return len;
+    }
+    /* For partial blocks CS2 is equivalent to CS3 */
+    return cts128_cs3_decrypt(ctx, in, out, len);
+}
+#endif
+
+int aes_cbc_cts_block_update(void *vctx, unsigned char *out, size_t *outl,
+                             size_t outsize, const unsigned char *in,
+                             size_t inl)
+{
+    PROV_CIPHER_CTX *ctx = (PROV_CIPHER_CTX *)vctx;
+    size_t sz = 0;
+
+    if (inl < 16) /* There must be at least one block for CTS mode */
+        return 0;
+    if (outsize < inl)
+        return 0;
+    if (out == NULL) {
+        *outl = inl;
+        return 1;
+    }
+
+    /*
+     * Return an error if the update is called multiple times, only one shot
+     * is supported. We use the bufsz to mark the update.
+     */
+    if (ctx->bufsz > 0)
```

Comment:
Surely a different flag variable would be clearer?

Reply:
I used another bit flag..
The ctx->flags is probably not a good idea.